app/gui/view/debug_scene/cloud-dashboard/src/projects_spinner.rs

```diff
@@ -0,0 +1,712 @@
+    }
+
+    fn init_color_and_visibility(
+        &self,
+        app: &application::Application,
+    ) -> stream::WeakNode<frp::SourceData> {
+        let scene = &app.display.default_scene;
+        let styles = StyleWatchFrp::new(&scene.style_sheet);
+        let frp = &self.frp;
+        let model = &self.model;
+        let network = &frp.network;
+        let hover = &model.hover;
+        let ring = &model.ring;
+        let spinner = &model.spinner;
+        let play = &model.play;
+        let pause = &model.pause;
+
+        frp::extend! { network
+            use theme::cloud_dashboard::play_button;
+            let non_toggled_color       = styles.get_color(play_button::non_toggled);
+            let toggled_color           = styles.get_color(play_button::toggled);
+            let hovered_color           = styles.get_color(play_button::hovered);
+            let toggled_hovered_color   = styles.get_color(play_button::toggled_hovered);
+            init_color_scheme           <- source::<()>();
+
+            hovered <- bool(&hover.events.mouse_out, &hover.events.mouse_over);
+            pressed <- bool(&hover.events.mouse_up_primary, &hover.events.mouse_down_primary);
+            color <- all_with3(&init_color_scheme, &hovered, &pressed,
+                move |_cs, h, p| {
+                    match (h, p) {
+                        (false, false) => non_toggled_color.value(),
+                        (false, true)  => toggled_color.value(),
+                        (true, false)  => hovered_color.value(),
+                        (true, true)   => toggled_hovered_color.value(),
+                    }
+                });
+            color_rgb <- color.map(|c| Vector3(c.red, c.green, c.blue));
+            eval color_rgb ((color) ring.color_rgb.set(*color));
+            eval color_rgb ((color) spinner.color_rgb.set(*color));
+            eval color ((color) play.set_color(*color));
+            eval color ((color) pause.set_color(*color));
+
+            spinner_visible <- frp.state.map(State::spinner_visible);
+            ring_visible <- spinner_visible.not();
+
+            eval spinner_visible ((visible) spinner.alpha.set(if *visible { 1.0 } else { 0.0 }));
+            eval ring_visible ((visible) ring.alpha.set(if *visible { 1.0 } else { 0.0 }));
+        }
+
+        init_color_scheme
+    }
+
+    fn init_visibility(&self) {
+        let frp = &self.frp;
+        let model = &self.model;
+        let network = &frp.network;
+        let ring = &model.ring;
+        let spinner = &model.spinner;
+
+        frp::extend! { network
+            spinner_visible <- frp.state.map(State::spinner_visible);
+            ring_visible <- spinner_visible.not();
+
+            eval spinner_visible ((visible) spinner.alpha.set(if *visible { VISIBLE } else { INVISIBLE }));
+            eval ring_visible ((visible) ring.alpha.set(if *visible { VISIBLE } else { INVISIBLE }));
+        }
+    }
+
+    fn init_event_tracing(&self) {
```

Comment:
Have you checked out the TRACE_ALL clause? You can add it to every extend! and define_endpoints_2!. See https://github.com/enso-org/enso/blob/e249f0a2ce73ccf42340a6f7b82c5d9aed334954/lib/rust/frp/src/macros.rs#L18

Reply:
No I had not seen this, obviates this function entirely, thank you 😄